### PR TITLE
Use ~~dataclasses~~ attrs for pt.Array

### PR DESCRIPTION
**Pull request by kaushikcfd**
Pros:
- Immutable types (`frozen=True`)
- Derived classes can be dataclasses (noticed while implementing #364)
- In almost all cases uses the generated `__init__`.

Cons:
- All derived classes must do `eq=False` in their call to `dataclass`, otherwise `dataclasses` will implement its own default which has a complexity of $O(2^N)$ in number of nodes in the DAG.
- All derived classes must do `def __hash__(self) -> int: return super().__hash__()`, otherwise `dataclasses` will implement its own default.
- Constructor arguments have to be `_shape`, `_dtype` instead of `shape`, `dtype` as `pt.Array` already has properties of that name.
- :warning: Pylint Failures.

### CODE REVIEW

**Comment by inducer** [comment]
This generally seems worthwhile, I think.

>     * All derived classes must do `eq=False` in their call to `dataclass`, otherwise `dataclasses` will implement its own default which has a complexity of O(2N) in number of nodes in the DAG.
>     * All derived classes must do `def __hash__(self) -> int: return super().__hash__()`, otherwise `dataclasses` will implement its own default.

Is this something we could check for in `__post_init__` in `__debug__` mode?

**Comment by kaushikcfd** [comment]
> Ah well. Maybe just disable unexpected-keyword-arg for pylint?

Yep, I will do that for now. Related pylint issue: https://github.com/PyCQA/pylint/issues/7623.

**Comment by kaushikcfd** [comment]
The pylint errors have been ignored. But the current refactoring forces us to depend on `>=Py3.10` as the implementation depends on [kw_only](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) argument of `dataclasses.field` that was introduced in 3.10. Maybe we can merge this once we force the requirement to Py3.10?

**Comment by kaushikcfd** [comment]
> Is this something we could check for in __post_init__ in __debug__ mode?

Yes, I think so. I will try to give it a shot.

**Comment by inducer** [comment]
> The pylint errors have been ignored. But the current refactoring depends on [kw_only](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) argument of `dataclasses.field` that was introduced in 3.10. Maybe we can merge this once we force the requirement to Py3.10?

Is there a way to avoid that? Requiring 3.10 is a healthy ways down the road in my view, given that we're not even completely switched over to 3.8.

**Comment by kaushikcfd** [comment]
> Is there a way to avoid that?

We could use [`attrs`](https://www.attrs.org/en/stable/examples.html) instead of `dataclasses`?

**Comment by inducer** [comment]
> We could use [`attrs`](https://www.attrs.org/en/stable/examples.html) instead of `dataclasses`?

As long as it's easy enough to switch back to dataclasses, I would not be oppposed.

**Comment by kaushikcfd** [comment]
> As long as it's easy enough to switch back to dataclasses, I would not be opposed.

Sure! It should take us an afternoon to switch to dataclasses from `attrs`.

File: setup.py [hunk 1]
```diff
@@ -38,6 +38,7 @@
         "pytools>=2021.1",
         "pyrsistent",
         "immutables",
+        "attrs",
```
Comment (inducer) [inline setup.py:41]: We should add a design decision in the docs explaining why we use `attrs` and not `dataclasses`.

File: pytato/array.py [hunk 2]
```diff
@@ -428,16 +434,17 @@ class Array(Taggable):
...
     # hashable. Dicts of hashable keys and values are also permitted.
     _fields: ClassVar[Tuple[str, ...]] = ("axes", "tags",)
 
-    __array_priority__ = 1  # disallow numpy arithmetic to take precedence
-
-    def __init__(self, axes: AxesT, tags: FrozenSet[Tag]) -> None:
-        self.axes = axes
-        self.tags = tags
+    # disallow numpy arithmetic from taking precedence
+    __array_priority__: ClassVar[int] = 1
```
Comment (inducer) [inline pytato/array.py:447]: During review, we flagged a risk that subclasses might use `eq=True` or use the default hash by accident, and that we wanted to catch this during `__post_init__`. Somehow, this must have fallen through the cracks, but I think we still ought to do that.

File: setup.py (hunk 1 above)
Comment (kaushikcfd) [reply setup.py:41]: Moving this discussion to https://github.com/inducer/pytato/pull/371.

File: pytato/array.py (hunk 2 above)
Comment (kaushikcfd) [reply pytato/array.py:447]: Moving this discussion to https://github.com/inducer/pytato/pull/371.

**Pull request merged**